File: algorithms/javaalgo/src/test/java/leetcode/number701to800/number701to710/number704/BinarySearchTest.java
package leetcode.number701to800.number701to710.number704;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class BinarySearchTest {

    private int[] nums;
    private int target;

    @Test
    void test() {
        int[] nums = {-1, 0, 3, 5, 9, 12};
        int target = 9;

        assertThat(search(nums, target)).isEqualTo(4);
        assertThat(forSearch(nums, target)).isEqualTo(4);
    }

    public int search(int[] nums, int target) {
        this.nums = nums;
        this.target = target;
        return dfs(0, nums.length - 1);
    }

    private int dfs(int left, int right) {
        if (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] < target) {
                return dfs(mid + 1, right);
            } else if (nums[mid] > target) {
                return dfs(left, mid - 1);
            } else {
                return mid;
            }
        } else {
            return -1;
        }
    }

    public int forSearch(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] < target) {
                left++;
            } else if (nums[mid] > target) {
                right--;
            } else {
                return mid;
            }
        }
        return -1;
    }
}

File: algorithms/javascriptalgo/leetcode/number701-800/number701-710/number704/binarySearch.test.js
test("test", () => {
  const nums = [-1, 0, 3, 5, 9, 12];
  const target1 = 9;
  const target2 = 2;

  expect(search(nums, target1)).toEqual(4);
  expect(search(nums, target2)).toEqual(-1);
});

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
const search = function (nums, target) {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor(left + (right - left) / 2);
    if (nums[mid] < target) {
      left++;
    } else if (nums[mid] > target) {
      right--;
    } else {
      return mid;
    }
  }
  return -1;
};
